homeassistant/components/camera/__init__.py

```diff
@@ -80,33 +82,59 @@ def camera_image(self):
         """Return bytes of camera image."""
         raise NotImplementedError()
 
-    def mjpeg_stream(self, response):
-        """Generate an HTTP MJPEG stream from camera images."""
-        def stream():
-            """Stream images as mjpeg stream."""
-            try:
-                last_image = None
-                while True:
-                    img_bytes = self.camera_image()
-
-                    if img_bytes is not None and img_bytes != last_image:
-                        yield bytes(
-                            '--jpegboundary\r\n'
-                            'Content-Type: image/jpeg\r\n'
-                            'Content-Length: {}\r\n\r\n'.format(
-                                len(img_bytes)), 'utf-8') + img_bytes + b'\r\n'
-
-                        last_image = img_bytes
-
-                    time.sleep(0.5)
-            except GeneratorExit:
-                pass
-
-        return response(
-            stream(),
-            content_type=('multipart/x-mixed-replace; '
-                          'boundary=--jpegboundary')
-        )
+    @asyncio.coroutine
+    def async_camera_image(self):
+        """Return bytes of camera image.
+
+        This method must be run in the event loop.
+        """
+        image = yield from self.hass.loop.run_in_executor(
+            None, self.camera_image)
+        return image
+
+    @asyncio.coroutine
+    def handle_async_mjpeg_stream(self, request):
+        """Generate an HTTP MJPEG stream from camera images.
+
+        This method must be run in the event loop.
+        """
+        response = web.StreamResponse()
+
+        response.content_type = ('multipart/x-mixed-replace; '
+                                 'boundary=--jpegboundary')
+        response.enable_chunked_encoding()
+        yield from response.prepare(request)
+
+        def write(img_bytes):
+            """Write image to stream."""
+            response.write(bytes(
```

Comment:
It's nonblocking: http://aiohttp.readthedocs.io/en/stable/web_reference.html#aiohttp.web.StreamResponse.write

Instead, we `yield from response.drain()`